### PR TITLE
Implement nonzero attack_weight, closes #3923

### LINKED ISSUE #3923: attack_weight does not function as advertised
[Reported on the forums](https://forums.wesnoth.org/viewtopic.php?p=638187#p638187)

Attack and defense weight are treated as a boolean setting rather than a true weighting.

### PR DESCRIPTION
**Pull request by ProditorMagnus**

### CODE REVIEW

**Comment by ProditorMagnus** [comment]
I assume it wont compile, probably needs -> somewhere

File: src/actions/attack.cpp [hunk 1]
```diff
@@ -502,6 +502,11 @@ bool battle_context::better_combat(const combatant& us_a,
 	a = (us_a.average_hp() - poison_a_us) * harm_weight - (them_a.average_hp() - poison_a_them);
 	b = (us_b.average_hp() - poison_b_us) * harm_weight - (them_b.average_hp() - poison_b_them);
 
+	double attack_weight_a = a.u_.weapon.attack_weight();
+	double attack_weight_b = b.u_.weapon.attack_weight();
+	a = a * attack_weight_a;
```
Comment (ProditorMagnus) [inline src/actions/attack.cpp:507]: probably should be done together with harm_weight

**Comment by gfgtdf** [comment]
im thinking it might be better to implement attack_weight as a weapon special instead, that way the weight depends on certain stuff, in particular to increae the weight at the same time as when other abilities are active.

**Comment by ProditorMagnus** [comment]
That makes sense, but unless you plan to do that in next few months its still worth to fix the intended use case.

**Comment by ProditorMagnus** [comment]
Would be interesting to know if the change applies for player, ai, or both.

**Comment by CelticMinstrel** [comment]
I'm not even sure _what_ this does, to be honest. Logically speaking, the attack weight is primarily for AI, but this code doesn't look like AI code. It seems to be the attack simulation code though, so it likely affects the AI.

**Comment by ProditorMagnus** [comment]
As purpose of attack_weight I expect
1) ai attack selection
2) player preselected attack

File: src/actions/attack.cpp [hunk 2]
```diff
@@ -498,9 +498,12 @@ bool battle_context::better_combat(const combatant& us_a,
 	double poison_b_us = us_b.poisoned > 0 ? (us_b.poisoned - us_b.hp_dist[0]) * game_config::poison_amount : 0;
 	double poison_b_them = them_b.poisoned > 0 ? (them_b.poisoned - them_b.hp_dist[0]) * game_config::poison_amount : 0;
 
+	double attack_weight_a = us_a.u_.weapon->attack_weight();
+	double attack_weight_b = us_b.u_.weapon->attack_weight();
+
 	// Compare: damage to them - damage to us (average_hp replaces -damage)
```
Comment (ProditorMagnus) [inline src/actions/attack.cpp:504]: not sure how the formula works here, whether I put weight to correct place

**Comment by Pentarctagon** [comment]
@gfgtdf any issue merging this as-is?

**Comment by CelticMinstrel** [comment]
Is it tested? Does it work?

**Comment by ProditorMagnus** [comment]
As compiled change I cant manually test it before release. I do not know how to cover it in unit test either.

**Comment by Pentarctagon** [comment]
You can download an artifact from https://github.com/wesnoth/wesnoth/actions/runs/9929715710?pr=9096 and try it.

**Comment by CelticMinstrel** [comment]
It's definitely something that's not easy to cover in a unit test. I think the best you could do is have two attacks with weights of 90 and 10, run through a large number of attacks (say 100), and verify that the 90-weight attack is used more than twice as often as the 10-weight attack.

But even that isn't exactly easy when what you want to test is actually the AI. I don't even know of any unit tests that test AI?

**Comment by ProditorMagnus** [comment]
Both use cases fail.

**Comment by ProditorMagnus** [comment]
I guess issue is in math

a = 50 = 100 - 50
b = 90 = 100 - 10

a-b=-40, pick b, correct


a = -400 = 100 - 50*10
b = 90 = 100 - 10

a-b=-490, pick b, wrong

I guess I need to fit in them_a.average_hp() * attack_weight_a - them_b.average_hp() * attack_weight_b somewhere

**Comment by ProditorMagnus** [comment]
Maybe it just needs negating the weight

a = 600 = 100 - 50*(-10)
b = 90 = 100 - 10

a-b=510, pick a

**Comment by CelticMinstrel** [comment]
> Maybe it just needs negating the weight

Negating a weight doesn't sound like it would make sense, but `(1-weight)` could be sensible.

**Comment by ProditorMagnus** [comment]
I dont think this is possible without having access to damage. I could weight the difference of attack end results, but weight is supposed to be applied in the common damage too.

**Comment by ProditorMagnus** [comment]
Maybe found where to get damage.

**Comment by ProditorMagnus** [comment]
Test OK for both user and AI.

TC1: 5-4 + 3*4-4 https://i.vgy.me/qmxFQI.png
Additional result: If there is chance to kill, that is more important than damage according to existing logic.

TC2: 9-4 + 3*2-4 https://i.vgy.me/hBZKDl.png

TC3: 5-4 melee nonmarksman + 10*0.001-5 rangedhttps://i.vgy.me/QK2yE4.png
Additional result: If there is chance to kill, that is more important than damage according to existing logic.

**Pull request merged**